lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -77,45 +96,210 @@ def __init__(
+
+        if access is None and user is None:
+            # {}
+            if is_admin:
+                # ADMIN needs no authorization search terms: all is permitted,
+                # so we need only the subclass query terms.
+                query["bool"] = {"filter": must}
+            else:
+                if not authorized_user:
+                    # An unauthorized client sees all public data
+                    must.append({"term": {"authorization.access": "public"}})
+                    query["bool"] = {"filter": must}
+                else:
+                    # An authorized user can see their own private data plus
+                    # public data ("OR" relationship)
+                    must.append(
+                        {
+                            "bool": {
+                                "should": [
+                                    {
+                                        "term": {
+                                            "authorization.owner": str(
+                                                authorized_user.id
+                                            )
+                                        }
+                                    },
+                                    {"term": {"authorization.access": "public"}},
+                                ]
+                            }
+                        }
+                    )
+                    query["constant_score"] = {"filter": {"bool": {"must": must}}}
+        elif access is None:
+            # {"user": "xxx"}
+            must.append({"term": {"authorization.owner": user}})
+            query["bool"] = {"filter": must}
+        elif user is None:
+            # {"access": "xxx"}
+            if access == Dataset.PUBLIC_ACCESS:
+                must.append({"term": {"authorization.access": Dataset.PUBLIC_ACCESS}})
+            else:
+                if not authorized_user:
+                    # An unauthorized user can never view access:private data,
+                    # but schema validation wasn't able to detect this so we
+                    # raise our special exception to skip the query.
+                    raise NoSelectedDatasets(user, access)
+                must.append({"term": {"authorization.access": Dataset.PRIVATE_ACCESS}})
+                if not is_admin:
+                    # A non-admin user can only see their own private data
+                    must.append(
+                        {"term": {"authorization.owner": str(authorized_user.id)}}
+                    )
+            query["bool"] = {"filter": must}
         else:
-            term = {"authorization.access": Dataset.PUBLIC_ACCESS}
-        return term
+            # {"access": "xxx", "user": "xxx"}
+            must.append({"term": {"authorization.access": access}})
+            must.append({"term": {"authorization.owner": user}})
+            query["bool"] = {"filter": must}
+        return query
```

Comment:
Looking through it I don't see any obvious holes, but I also think this is one of those cases I prefer a not quite so extremely consolidated logic chain so it's easier to read the flow. But I had already intended to go back and think about whether it could be optimized; I'll feed this into the grinder.